Improve performance of dropmissing

svilupp opened the pull request:
Fixes https://github.com/JuliaData/DataFrames.jl/issues/3254

TODO: 
~~[x] change dropmissing~~
~~[x] change dropmissing! (no opportunities found)~~
~~[x] add/augment testing suite~~
~~[x] check docs and examples for consistency~~

bkamins commented:
Please also make sure to review tests so that they adequately cover any new code paths that might be needed.
Regarding performance you might also want to have a look at `completecases` as there might be some room for performance gains also (something like e.g. `reduce_or!`) - but if indeed there is some performance gain potential would need to be benchmarked.

svilupp commented:
Yes, of course. I'll review/add tests as necessary.

I have quickly reviewed how it works now and I'm surprised you expect benefits from fusing the copy rows into `newdf` step and `disallowmissing`. I'll study it a bit more and benchmark some variants.

Re. `completecases`, I was actually wondering if it all could be done in a single pass without too much additional code/complexity, eg,
- initiate all vectors with undef
- iterate by rows and if `!ismissing` copy them in
- at the end "shrink" the vectors if some slots were left empty (because there were missing)

Thanks for the tip! Let me know if you have any more tips or ideas. I'll get into it in a few days when I'm done travelling.
Is `reduce_or` an existing function or a suggestion for an approach? I couldn't find it.

bkamins commented:
Your comments are very good. I just threw some ideas that I had. Unfortunately, as you say, it is often the case that one needs to run benchmarks before one can find the best approach as things sometimes are surprising. But for sure there is room for improvement, as DuckDB is noticeably faster.

Two additional notes:
* it might be the case that for small number and large number of rows other strategies will be efficient
* similarly it might be the case that number of columns in a data frame matters (and here are two dimensions: total number of columns and number of columns inspected for missing values).

svilupp commented:
**Quick update from my side**
- build a benchmarking suite to test performance across different length/width/missingness share
- unfortunately, learned that my single pass idea is 1-2 orders of magnitude slower
- main driver -> cannot achieve indexing into a DataFrame without allocations (I suspect the unstable indexing return type)
- not sure that single pass method is possible with many columns (as we need to go "horizontally")

**Single-pass Implementation**
- I chose the following base case: 10k rows (`row_total`), 10 columns (`cols_total`), 3 columns with missing data (`cols_missing`), 10% missingness (`missingness_share`)
- part of the slowdown is that my function checks all columns for missingness vs the default can skip some (kwargs and optional subsets can be added later.. ) Perhaps a fairer benchmark would be with `cols_total == `cols_missing`, but the current issue is that ->
- I'm getting a ridiculous amount of allocations and am unsure how to avoid them (it seems to come from the "instability" of DataFrame indexing across many columns)

```
# the simplest case with namedtuple row iterator, very slow
function dropmissing2(df::DataFrame)
    # init all vectors
    newdf = DataFrame()
    for col in names(df)
        newdf[!, col] = Vector{Union{eltype(df[!, col]),Missing}}(undef, nrow(df))
    end
    # iterate by row
    last_row = 0
    for row in DataFrames.Tables.namedtupleiterator(df)
        # none of the values are missing
        if !any(ismissing, values(row))
            last_row += 1
            for (key, val) in pairs(row)
                newdf[last_row, key] = val
                # newdf[!, key][idx] = val
            end
        end
    end
    # resize
    resize!(newdf, last_row)
    newdf
end

# construct with pairs
init_without_missing(df) = DataFrame([col => Vector{nonmissingtype(eltype(df[!, col]))}(undef, nrow(df)) for col in names(df)]...)

# construct via internal storage
init_without_missing2(df, nrows, cols) = DataFrame([Vector{nonmissingtype(eltype(col))}(undef, nrows) for col in DataFrames._columns(df)], cols)

# for-loop iteration directly on vectors
function dropmissing3(df::DataFrame)
    nrows = nrow(df) # 2 alloc
    cols = names(df) # 11 alloc
    # init all vectors
    newdf = init_without_missing2(df, nrows, cols) # 90 alloc
    # iterate by row
    last_row = 1
    @inbounds for i in 1:nrows
        nomissingfound = true
        for col in cols
            nomissingfound *= !ismissing(df[i, col]) # 1 alloc for each access?
            # stop if missing is found
            !nomissingfound && break
            newdf[last_row, col] = df[i, col] # 1 alloc for each access?
        end
        # if there was a missing, overwrite this row in the next iteration
        if nomissingfound
            last_row += 1 # increment for next pass
        end
    end
    # resize to required size
    resize!(newdf, last_row - 1) # 10 alloc
    newdf
end

@time newdf = dropmissing3(df);
@time newdf = dropmissing3(df);
# 0.045442 seconds (517.22 k allocations: 9.420 MiB)

# correctness check
isequal.(Matrix(newdf), Matrix(dropmissing(df))) |> all
# true
```
Benchmarking results
```
# benchmark new implementations
basecase = (; rows_total=10000, cols_total=10, cols_missing=3, missingness_share=0.1)
pl = run_benchmark_suite(Val(:belapsed), [dropmissing, dropmissing!, dropmissing3], basecase)
```
<img width="627" alt="image" src="https://user-images.githubusercontent.com/49557684/209593279-944c5725-99bc-43b1-b113-a929dc135051.png">


**Benchmarking setup**
```
using DataFrames
using BenchmarkTools
using Plots

"generate a data frame with missing values for various scenarios"
function generate_data(; rows_total, cols_total, cols_missing, missingness_share)
    data = rand(rows_total, max(cols_total, cols_missing))
    data = Matrix{Union{Float64,Missing}}(data)
    for j in axes(data, 2)
        if j <= cols_missing
            data[:, j] .= ifelse.(@view(data[:, j]) .<= missingness_share, missing, @view data[:, j])
        end
    end
    DataFrame(data, :auto)
end

"time the function execution on the second run"
timer(method::Val{:elapsed}, func, bc; kwargs...) = (df = generate_data(; bc..., kwargs...); func(df); @elapsed func(df))
timer(method::Val{:belapsed}, func, bc; kwargs...) = (df = generate_data(; bc..., kwargs...); @belapsed($func($df)))
timer(method::Val{:allocated}, func, bc; kwargs...) = (df = generate_data(; bc..., kwargs...); func(df); @allocated func(df))

nt(key, val) = NamedTuple{(Symbol(key),)}((val,))
val_to_title(m) = match(r":([a-z]+)", string(m)).captures[1] |> uppercase

"plot scenarios in place"
function plot_scenarios!(pl, method, func::Function, bc; kwargs...)
    tested_key = intersect(keys(bc), keys(kwargs)) |> only
    data = [timer(method, func, bc; nt(tested_key, val)...) for val in kwargs[tested_key]]
    title = "Effect of $(uppercase(string(tested_key))) on $(val_to_title(method))"
    plot!(pl, kwargs[tested_key], data; label=string(func),
        title, xaxis=:log, yaxis=:log)
    return pl, data
end
function plot_scenarios(method, func::Function, bc; kwargs...)
    pl, data = plot_scenarios!(plot(), method, func, bc; kwargs...)
end

"plot scenarios with multiple functions"
function plot_scenarios(method, func_vec::AbstractVector, bc; pl=plot(), kwargs...)
    for func in func_vec
        pl, _ = plot_scenarios!(pl, method, func, bc; kwargs...)
    end
    pl, nothing
end

"Runs benchmarking and plots results in a 2x2 grid"
function run_benchmark_suite(method, func, bc)
    pltr(; kwargs...) = plot_scenarios(method, func, bc; kwargs...)
    plot([pltr(; rows_total=[10^3, 10^4, 10^5, 10^6])[1],
            pltr(; cols_total=[10, 100, 1000])[1],
            pltr(; cols_missing=[10, 100, 1000])[1],
            pltr(; missingness_share=[0.0001, 0.001, 0.01, 0.1])[1]]...,
        plot_title=string("DataFrames `dropmissing` Benchmark"),
        plot_titlefontsize=12,
        titlefontsize=10, size=(800, 600))
end

# benchmark an individual run
m = Val(:elapsed)
basecase = (; rows_total=10000, cols_total=10, cols_missing=3, missingness_share=0.1)
df = generate_data(; basecase...)
timer(m, dropmissing, basecase) # time a single run

# benchmark along a single dimension
plot_scenarios(m, dropmissing, basecase; rows_total=[10^3, 10^4, 10^5, 10^6]) |> first

# multiple functions at once
plot_scenarios(m, [dropmissing, dropmissing!], basecase; rows_total=[10^3, 10^4, 10^5, 10^6]) |> first

# execute the benchmarking suite
basecase = (; rows_total=10000, cols_total=10, cols_missing=3, missingness_share=0.1)
pl = run_benchmark_suite(Val(:elapsed), [dropmissing, dropmissing!], basecase)
```

bkamins commented:
`newdf[last_row, key] = val` for sure will be slow as it is type unstable.

In general 10k rows is small. Of course it would be good to optimize also for such case, but I would say priority is to optimize against 10^6+ rows.

Maybe a good idea would be to start with profilling `dropmissing` and `dropmissing!` to see where the slowest parts are?

svilupp commented:
> In general 10k rows is small. 

Noted! I've moved up the benchmark to from 10k to 10M.

**Fusing together dropmissing + disallowmissing**
```
# construct via internal storage
init_without_missing3(df, nrows) = DataFrame(AbstractVector[Vector{nonmissingtype(eltype(col))}(undef, nrows) for col in DataFrames._columns(df)], copy(DataFrames.index(df)), copycols=false)

# no-copy filler with binary mask (the hope was to be specialized and faster)
function fill_with_mask!(output_vec, input_vec, copy_mask)
    last_row = 0
    @inbounds @simd for i in eachindex(input_vec, copy_mask)
        if copy_mask[i]
            last_row += 1
            output_vec[last_row] = input_vec[i]
        end
    end
end

function dropmissing6(df::AbstractDataFrame)
    rowidxs = completecases(df)
    # init without missing
    newdf = init_without_missing3(df, sum(rowidxs))
    # grab data references
    newdf_data = DataFrames._columns(newdf)
    df_data = DataFrames._columns(df)
    for i in eachindex(newdf_data, df_data)
        # fill data references without copying
        fill_with_mask!(newdf_data[i], df_data[i], rowidxs)
    end
    return newdf
end

### Alternative
# getter without missing type
function getindex_not_missing(vec, missing_mask)
    output = Vector{nonmissingtype(eltype(vec))}(undef, sum(missing_mask))
    output[:] .= vec[missing_mask]
    output
end

function dropmissing7(df::AbstractDataFrame)
    rowidxs = completecases(df)
    # init without missing
    new_columns = Vector{AbstractVector}(undef, ncol(df))
    # grab data references
    df_data = DataFrames._columns(df)
    for i in eachindex(new_columns, df_data)
        @inbounds new_columns[i] = getindex_not_missing(df_data[i], rowidxs)
    end
    return DataFrame(new_columns, copy(DataFrames.index(df)), copycols=false)
end
```

**Results**
```
# Benchmark - `dropmissing` still wins in benchmarks
@time dropmissing(df); # default
# 0.013654 seconds (129 allocations: 59.327 MiB)
@time dropmissing6(df); # fewer allocs but slower
# 0.035927 seconds (106 allocations: 26.819 MiB)
@time a = dropmissing7(df); # best trade-off but not fast enough
# 0.016576 seconds (116 allocations: 56.673 MiB)

# Correctness check - PASSES
@time newdf = dropmissing6(df);
isequal.(Matrix(newdf), Matrix(dropmissing(df))) |> all
# true
@time newdf = dropmissing7(df);
isequal.(Matrix(newdf), Matrix(dropmissing(df))) |> all
# true
```

<img width="629" alt="image" src="https://user-images.githubusercontent.com/49557684/209685861-00d37294-0a88-4254-99ff-ea3f0a837e5f.png">
The default algorithm is still pulling ahead in most measures (especially in larger vectors, which is unexpected with fewer allocations needed...)


**Zoom-in on allocations**
<img width="627" alt="image" src="https://user-images.githubusercontent.com/49557684/209686205-a8006f64-50c3-424c-8c38-12c93d97cad5.png">
Clearly, variant 6 has fewer allocations but is not faster

svilupp commented:
Finally some progress, but I'm struggling to get more than 2x speedup (results below).
<img width="1009" alt="image" src="https://user-images.githubusercontent.com/49557684/211173812-1bcab7e2-5692-452d-8217-de9f289ddc19.png">

Changes:
- keep completecases bitvector but convert it to the indices of non-missing rows at the top level
- then we can allocate the right sized vector without missing type and fill them (`getindex_not_missing3`)
- threading added to honour the same rules as the default implementation

**Setup**
```
"Getter function that produces a vector with missing type"
function getindex_not_missing3(vect, rowidxs)
    output = Vector{nonmissingtype(eltype(vect))}(undef, length(rowidxs))
    last_ind = 0
    @inbounds for i in eachindex(rowidxs)
        output[last_ind+=1] = vect[rowidxs[i]]
    end
    output
end
vect = ones(Union{Float32,Missing}, 10^7)
row_pos = 1:10^7
@time getindex_not_missing3(vect, row_pos) # 2 allocations
# 0.008571 seconds (2 allocations: 38.147 MiB)

function dropmissing8(df::AbstractDataFrame)
    # Note: timings on the right hand side are for a vector of 10^7 Float32 elements
    # get bitvector of rows with no missing values
    rowmask = completecases(df) # 8.6ms
    # convert to indices -> this also informs us about the right size
    rowidxs = DataFrames._findall(rowmask) # 5-6ms
    # init new columns (data)
    new_columns = Vector{AbstractVector}(undef, ncol(df))
    # grab data references
    df_columns = DataFrames._columns(df)
    # threading decision rule borrowed from `_threaded_getindex`
    if nrow(df) >= 1_000_000 && Threads.nthreads() > 1
        @sync for i in eachindex(new_columns)
            # creates a vector of the right size, without missing type, and copies only data at indices in rowidxs
            # replaces filtering + disallowmissing in two steps
            Threads.@spawn @inbounds new_columns[i] = getindex_not_missing3(df_columns[i], rowidxs) #c. 3.6ms per call
        end
    else
        for i in eachindex(new_columns)
            @inbounds new_columns[i] = getindex_not_missing3(df_columns[i], rowidxs)
        end
    end
    return DataFrame(new_columns, copy(DataFrames.index(df)), copycols=false)
end
```

**Timings**
```
basecase = (; rows_total=10000000, cols_total=10, cols_missing=10, missingness_share=0.1)
df = generate_data(; basecase...)
@time dropmissing8(df);
# 0.036446 seconds (270 allocations: 295.067 MiB)
@btime dropmissing8($df); # 51-52ms not threaded / 27-28ms threaded
# 27.359 ms (172 allocations: 295.06 MiB)

# Baseline
@time dropmissing(df);
# 0.074098 seconds (205 allocations: 594.339 MiB)
@btime dropmissing($df);
# 60.561 ms (202 allocations: 594.34 MiB)

# Correctness test - pass
@time newdf = dropmissing8(df);
isequal.(Matrix(newdf), Matrix(dropmissing(df))) |> all # true
```

svilupp commented:
Just as a curiosity - I was curious how it would look if Threading was always on (`dropmissing9`)
- It seems that on my system the break-even is at 1000 rows (with 10 columns/6 threads), whilst the decision rule looks for >1 million rows... 
- In addition, I'm not sure what the difference is between Threads.@threads and @spawn-ing, but the former seems to have better performance on my system? (I need to double check it's not because of removing the IF condition, but that should be as costly)

<img width="592" alt="image" src="https://user-images.githubusercontent.com/49557684/211174017-c2710456-0d9f-46e6-8cf1-3d66296886e3.png">

```
"Identical to dropmissing8 but use `Threads.@threads`"
function dropmissing9(df::AbstractDataFrame)
    # Note: timings on the right hand side are for a vector of 10^7 Float32 elements
    # get bitvector of rows with no missing values
    rowmask = completecases(df) # 8.6ms
    # convert to indices -> this also informs us about the right size
    rowidxs = DataFrames._findall(rowmask) # 5-6ms
    #markinit new columns (data)
    new_columns = Vector{AbstractVector}(undef, ncol(df))
    # grab data references
    df_columns = DataFrames._columns(df)
    Threads.@threads for i in eachindex(new_columns)
        # creates a vector of the right size, without missing type, and copies only data at indices in rowidxs
        # replaces filtering + disallowmissing in two steps
        @inbounds new_columns[i] = getindex_not_missing3(df_columns[i], rowidxs) #c. 3.6ms per call
    end
    return DataFrame(new_columns, copy(DataFrames.index(df)), copycols=false)
end
```

bkamins commented:
> It seems that on my system the break-even is at 1000 rows (with 10 columns/6 threads), whilst the decision rule looks for >1 million rows...

We wanted to be on a safe side (i.e. do not use threads unless there is a clear benefit). In particular - threading benefit might vary across machines on which code is run.

> In addition, I'm not sure what the difference is between Threads.@threads and @spawn-ing, but the former seems to have better performance on my system? (I need to double check it's not because of removing the IF condition, but that should be as costly)

We need to keep Julia 1.6 compatibility, where `@threads` only had static scheduler and there was an issue with `@threads` composability (i.e. when the DataFrames.jl operation that is potentially multi-threaded would be itself spawned in a multi-threaded code). `@spawn` does not have these issues.

svilupp commented:
Hi @bkamins,

I've tried to integrate the new implementation of dropmissing into the DF codebase. 
_Could you please check if its acceptable?_

**My thinking:**
- I tried to mimic your current design and variable names, but, unfortunately, I couldn't leverage the getindex methods
- This implementation mostly relies on `_getindex_disallowmissing` for a vector, which is meant to be an internal method fusing getindex+disallowmissing! together to save cycles; This is then wrapped in a method of the same name for DataFrames that handles when to choose `getindex()` vs `_getindex_disallowmissing` based on user's request
- While `getindex()` has many upstream use cases, this has only one, so I chose to write fewer lines of code (eg, not create separate methods for `dropmissing(df,:)` vs `dropmissing(df,:x1)` vs `dropmissing(df,[:x1,:x2])`)
- I've broken your abstraction hierarchy, where the special threading case is hidden in a dedicated function (`_getindex_threaded`) to avoid unnecessary LoC, and add it inside the loop of `getindex_disallowmissing`)

**Sharp edges:**
- Given the single purpose, I didn't add try-catch loops to `_getindex_disallowmissing` to give nicer errors, because it's an internal method and it's usage is safe (we know that those indices don't have `missing` thanks to `completecases`

I've tested correctness against the default function, but once we agree on the design, I'll add some individual tests.

**Next steps:**
[x] Docstrings. I've added comments to explain the logic, but its usage and implementation are not changing, so I intend to keep the current docstrings
[ ] Design. Agree on design/integration of the new function
[ ] Propose a new threading rule (num_rows)
[ ] Tests

**Side observations:**
- At large row volumes, mutating version `dropmissing!` gets very slow (my benchmark above was wrong as it didn't have `evals=1` in the benchmarking macro). I've checked the implementation with deleteat! and I don't see how I would significantly improve that without any allocation to keep it's non-mutating "spirit"
- I have re-ordered the threading if condition ("to thread or not to thread") to first check if there are any threads available (it's the cheapest one, so it can short-circuit quickly)


**Performance results:**
The new implementation does speed it up a little bit (`_dropmissing` is the one)
```
basecase = (; rows_total=10^7, cols_total=10, cols_missing=3, missingness_share=0.1)
df = generate_data(; basecase...)

# Perf
@time dropmissing(df);
# 0.176595 seconds (301 allocations: 594.392 MiB, 57.60% gc time)
@time _dropmissing(df);
# 0.078146 seconds (172 allocations: 295.083 MiB)
@btime dropmissing($df);
# 60.178 ms (202 allocations: 594.39 MiB)
@btime _dropmissing($df);
# 28.326 ms (172 allocations: 295.08 MiB)

# Correctness check
newdf = _dropmissing(df);
@assert isequal(newdf, dropmissing(df))
newdf = _dropmissing(df,[:x1,:x2]);
@assert isequal(newdf, dropmissing(df,[:x1,:x2]))
```

![benchmark_20230114](https://user-images.githubusercontent.com/49557684/212489685-da855509-4327-4425-918b-141f3ca249b3.png)

svilupp commented:
As for the decision of when to engage in threading, I did a quick benchmark (see below).
**My recommendation would be to start threading when `nrow(df)>100_000` (conservative)** 
However, I'd like to encourage consideration for a 10_000 threshold (IMO it's a better trade-off for real-world data sets).

I've looked at the breakeven point for threading across a different number of rows and number of columns (max 6 columns as I have 6 threads) and also with the special case when only 1 column has missings (ie, fewer rows dropped)

**Findings**:
- The breakeven on my machine seems to be around 50_000 rows already with 2 columns
- With 6 columns (leveraging all threads), this breakeven moves to 5_000 (with only 1 missing column, it's closer to 50k if all columns have missing; it's because more rows get dropped with more missing columns and hence the threading overhead is more significant)
- This breakeven keeps going down but seems to stay above 1_000 rows, eg, with 18 columns and 1 missing (turning around each thread 3 times), this breakeven moves to c. 2_000 rows
- I recommended 100_000 rows, because I suspect that threading might be more efficient on my machine (M1 Pro/32gb) vs the average user
- However, it is worth consideration to set it to ~10_000 rows --> assumption being that most users have 4 threads and most dataframes have >4 columns (ie, they are not in the "long" format during the clean-up phase); 
even if threading was slower, we're around `10^-4s` time scale, so a user will never notice any accidental slow down, but they significantly benefit if there are many columns (eg, with 18columns and 50k rows the speed-up from threading is 5x)

**The case when all columns have `missing` (more rows dropped)**
_Look where the green line (threaded code) crosses the red line (no threading)_
![benchmark_threading_missing_match_20230114](https://user-images.githubusercontent.com/49557684/212491374-3e32ffe9-ea88-4cde-9548-d9b7e14c8522.png)

**The case when only 1 column has a missing (fewer rows dropped)**
![benchmark_threading_missing_1_20230114](https://user-images.githubusercontent.com/49557684/212491673-f86755ab-b3ad-4eb4-94b7-4f8563997e2b.png)

**Versioninfo** 
> System: 6 threads, 32GB RAM
> Julia Version 1.8.4
Commit 00177ebc4fc (2022-12-23 21:32 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.5.0)
  CPU: 8 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 6 virtual cores


**Code:**
```
plt_array=[]
for num_cols in 1:6
    num_cols_missing=1 # or num_cols
    basecase = (; rows_total=10^6, cols_total=num_cols, cols_missing=num_cols_missing, missingness_share=0.1)
    @time pl = plot_scenarios(Val(:belapsed), [dropmissing,_dropmissing,_dropmissing_threaded], basecase; 
        rows_total=[10^3,5*10^3,10^4,5*10^4,10^5],title="# Cols: $num_cols / # Missing: $num_cols_missing") |> first
    push!(plt_array, plot(pl, title = "# Cols: $num_cols / # Missing: $num_cols_missing"))
end
pl=plot(plt_array...,layout=(2,3),plot_title="Effect of threading on dropmissing",size=(1000, 800))
```

bkamins commented on src/dataframe/dataframe.jl:
```diff
@@ -625,6 +625,45 @@ Base.getindex(df::DataFrame, row_ind::Colon, col_inds::MultiColumnIndex) =
 
+# Gets a subset of a DataFrame by `row_inds` and disallows missing values in columns specified by `col_disallowmissing_inds` 
+# Called by dropmissing()
+function _getindex_disallowmissing(df::DataFrame, row_inds::AbstractVector{T},
```
what if `df` is not a `DataFrame`?

svilupp replied:
Excellent point! I've been too focused on the main case.

Would it make sense to just rely on each type's own methods and do a general fallback?

bkamins commented on src/other/utils.jl:
```diff
@@ -477,3 +477,16 @@ function _findall(B::BitVector)::Union{UnitRange{Int}, Vector{Int}}
+    for i in eachindex(inds)
+        output[last_ind+=1] = vect[inds[i]]
+    end
+    output
```
```suggestion
    return output
```

bkamins commented on src/other/utils.jl:
```diff
@@ -477,3 +477,16 @@ function _findall(B::BitVector)::Union{UnitRange{Int}, Vector{Int}}
+    output = Vector{nonmissingtype(T)}(undef, length(inds))
+    last_ind = 0
+    for i in eachindex(inds)
+        output[last_ind+=1] = vect[inds[i]]
```
better split out `last_ind += 1` as an operation before. Probably also `@inbounds` here might make sense.

svilupp replied:
Done. I've not added inbounds, as it was called in the calling function already, but I've added it here as well.

bkamins commented on src/other/utils.jl:
```diff
@@ -477,3 +477,16 @@ function _findall(B::BitVector)::Union{UnitRange{Int}, Vector{Int}}
+# unsafe because it does not check for `inds` values to be not missing (ensured by the caller)
+# Called by `_getindex_disallowmissing(df,...)`
+function _getindex_disallowmissing(vect::AbstractVector{T}, inds::AbstractVector{<:Integer}) where {T}
+    output = Vector{nonmissingtype(T)}(undef, length(inds))
```
`vect` does not have to be `Vector`. Commonly it will be `PooledVector` or `CategoricalVector`. Typically we use `Tables.allocatecolumn` to get such behavior.

Additionally we might special case "pooled" vectors to preserve their pools as this will be faster (but this is a wider issue so it can be skipped in this PR).

svilupp replied:
Excellent point. I wasn't aware of `Tables.allocatecolumn`, so thanks for the tip!

The penalty is incredibly high but at least it's general
```
vect=rand(1:10, 10^7) |>allowmissing
@btime output = Vector{nonmissingtype(eltype($vect))}(undef, 10^7);
# 31.397 μs (2 allocations: 38.15 MiB)
@btime Tables.allocatecolumn(eltype($vect),10^7);
# 2.007 ms (2 allocations: 47.68 MiB)

cvect=CategoricalArray(vect)
@btime output = Vector{nonmissingtype(eltype($vect))}(undef, 10^7);
# 20.340 μs (2 allocations: 76.29 MiB)
@btime Tables.allocatecolumn(eltype($vect),10^7);
# 3.684 ms (2 allocations: 85.83 MiB)
```

svilupp replied:
I've had a quick look if I could add a special case for pooled vectors, but I believe the best way is to actually provide a general fallback that will leverage each special type's own methods.

Quick benchmark: general fallback is much faster (I presume that the element-by-element access on overall type is too expensive.

```
function general_fallback(vect,inds)
    vect[inds]|>disallowmissing
end

vect=rand(1:10, 10^7) |>allowmissing
cvect=CategoricalArray(vect)
row_mask = rand(10^7) .<= 0.5
row_inds = DataFrames._findall(row_mask)

@btime x=_getindex_disallowmissing($cvect, $row_inds);
# 33.966 ms (5000926 allocations: 171.69 MiB)

@btime x=general_fallback($cvect, $row_inds);
# 6.407 ms (20 allocations: 38.16 MiB)

# and we don't seem to pay a penalty for using indices instead of a bitvector mask
@btime x=general_fallback($cvect, $row_mask);
# 8.980 ms (20 allocations: 38.16 MiB)
```

If we wanted to specialize on pooled vectors, I'd probably need some pointers
PooledArray
- it's a dependency, so we could specialize directly
- It's probably not a good idea to use internals (copy pool, getindex refs vector and update the rest)
- it might be possible to fallback to general getindex|>disallowmissing fairly easily

CategoricalArrays
- I couldn't find any trait or reference with which I could specialize
- I went through DataAPI but the best is default array which we have already via Tables.allocatecolumn
- Alternatively, I could add weakdeps (1.9 extensions functionality) and fallback with Requires, but that's probably too involved for such narrow use case

bkamins commented on src/other/utils.jl:
```diff
@@ -477,3 +477,22 @@ function _findall(B::BitVector)::Union{UnitRange{Int}, Vector{Int}}
+# Called by `_getindex_disallowmissing(df,...)`
+function _getindex_disallowmissing(vect::AbstractVector{T}, inds::AbstractVector{<:Integer}) where {T}
+    # General fallback, relies on custom vectors' own methods
+    return vect[inds]|>disallowmissing
```
Let us continue the discussion with this inner function as it is a core.
I have benchmarked your proposal against doing `disallowmissing(@view vect[inds])` and it seems that using `@view` is comparable to your faster option (while being also relatively fast for your fallback option).

Can you please check it?

(after we finalize the "inner method", we will go back to general design.

Thank you!

svilupp replied:
Of course, you're right -- that is much more elegant and much faster! I wish I didn't waste as much time on the for-loops :)

These are the benchmarking results - it scales pretty well and this is for numeric columns! 
(the above proposal is named `_getindex_disallowmissing_bk`)
<img width="996" alt="image" src="https://user-images.githubusercontent.com/49557684/213864199-9259cbb3-1329-48b9-9805-64f6ee3b7d40.png">

It performs better than the default for "pooled" arrays and it's within c. 10-15% of the numeric columns, which is great.

I went ahead and changed the PR to reflect this approach.

***
Would the next be to write a testset for the new inner function?

bkamins replied:
So now I think that what should be done is change the PR so that the new function is invoked and run by standard tests.

Two general comments:
1. The `disallowmissing(@view vect[inds])` is so short that I do not think we need a separate function for it.
2. It seems that we do not have to have a separate code path for `AbstractDataFrame`. All could go into one function but just in place where you use `_columns` for `DataFrame`, `collect(eachcol(...` can be used for other data frame types
3. In several places you use `@inbounds` but I do not think they are needed (i.e. they are correct, but they should not affect the performance - in such cases it is better to use safer constructs and not use `@inbounds`)

In summary - I would propose to clean up the PR so that the code has a final form (and then I will review it) and make sure it passes existing tests (of course it is OK to add new tests now also).

svilupp replied:
Noted! I have done the cleanup as you suggested and the tests PASSED.

Re 1) Agreed. The only potential benefit could have been to ensure handling of all column types and ensuring "proper" return type 

Re 2) Agreed. I've made the change, but it comes at the cost of varying levels of abstraction (all other conditional branches are fairly "light"). 
I've tried to simplify the nested IFs, but I'm not sure how to go any further to ensure proper behaviour (?threaded, ?disallowmissing

Re 3) Removed it. I'm not sure why it doesn't matter, because there is still indexing and hence bounds checks should be happening, but since there seems to be no penalty, I've removed it.

svilupp replied:
One random note, please have a look at pathway for dropmissing(df::SubDataFrame)

The implemented behaviour is that dropmissing(::AbstractDF) will always produce a DataFrames (if view=false).

I'm not sure how `disallowmissing!` was implemented for SubDataFrame (couldn't find the method), but practically from trying it out `dropmissing(::SubDataFrame) returned DataFrame. So practically there shouldn't be any change (if I got it right).

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,52 @@ julia> dropmissing(df, [:x, :y])
     else
-        newdf = df[rowidxs, :]
-        disallowmissing && disallowmissing!(newdf, cols)
+        # With exact indices we can skip a lot of iterations (as opposed to indexing via a mask)
```
This is OK. Maybe add a comment that this especially makes sense because we assume that typically we keep most of rows (only few are dropped).

svilupp replied:
I'm not sure that would justify it. It generally faster than indexing via a mask, but there is an overhead of the _findall call, so this approach is justified if:
- if you're dropping many rows (which can happen easily if you have missing across many columns... the % of rows to drop goes up very quickly with the "ANY missing")
- if you have many columns (which is extremely common)

So what really justifies the `_findall` is the fact that most users have multiple columns. I'll reflect in the note

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,52 @@ julia> dropmissing(df, [:x, :y])
+        # With exact indices we can skip a lot of iterations (as opposed to indexing via a mask)
+        selected_rows = _findall(rowidxs)
+        new_columns = Vector{AbstractVector}(undef, ncol(df))
+        df_columns = if typeof(df) == DataFrame 
+            _columns(df)
+        else 
+            # for SubDataFrame
+            collect(eachcol(df))
+        end
```
```suggestion
        df_columns = df isa DataFrame ? _columns(df) : collect(AbstractVector, eachcol(df))
```
(`AbstractVector` to reduce compilation, but it probably is a minor thing)

svilupp replied:
Updated!

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,52 @@ julia> dropmissing(df, [:x, :y])
+            @sync for i in eachindex(new_columns)
+                # for each column, check if disallowmissing should be applied
+                Threads.@spawn if disallowmissing && (i in cols_inds)
+                    new_columns[i] = Missings.disallowmissing(@view df_columns[i][selected_rows])
```
```suggestion
                    new_columns[i] = disallowmissing(view(df_columns[i], selected_rows))
```
Your code was OK, but in this case I felt using the `view` function is easier to understand, as you are doing double indexing, and it is not immediately obvious for a reader if it gets applied to first or second index.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,52 @@ julia> dropmissing(df, [:x, :y])
+        else
+            for i in eachindex(new_columns)
+                if disallowmissing && (i in cols_inds)
+                    new_columns[i] = Missings.disallowmissing(@view df_columns[i][selected_rows])
```
```suggestion
                    new_columns[i] = disallowmissing(view(df_columns[i], selected_rows))
```

svilupp replied:
Makes sense with one change -- I need to keep the explicit import from Missings.
The `Missings.disallowmissing` function clashes with the local scope keyword `disallowmissing`.

The alternatively would be to change the keyword name, but that change the user-facing API, which is completely unnecessary.

svilupp replied:
Could you check if the function works for you by changing `@view` to `view`?

When I run it with `@view`, all is good.
When I use `view()`, I get an error, as if `disallowmissing` variable and function name were clashing:
(this is the exact error you'd get if you remove the explicit import of function `disallowmissing()`, because the function scope will override it)

> ERROR: MethodError: objects of type Bool are not callable
Maybe you forgot to use an operator such as *, ^, %, / etc. ?
Stacktrace:
  [1] #dropmissing#1056
    @ ~/Documents/GitHub/DataFrames.jl/src/abstractdataframe/abstractdataframe.jl:1005 [inlined]
  [2] dropmissing (repeats 2 times)
    @ ~/Documents/GitHub/DataFrames.jl/src/abstractdataframe/abstractdataframe.jl:970 [inlined]

It SHOULD work - I don't see why not and it works fine in smaller snippets I created trying to debug it / reproduce it (see f1 vs f2 below). 
The only context it fails in is within the `dropmissing` function.

_Can you make this change on your machine and run `dropmissing(df)` ?_


Failed attempts to reproduce the bug. 
LLVM is identical, so I wonder if it's something on the Julia parsing level (issue with references references)
```
function f1(df,selected_rows;disallowmissing=true)
    Missings.disallowmissing(@view(DataFrames._columns(df)[1][selected_rows]))
end
function f2(df,selected_rows;disallowmissing=true)
    Missings.disallowmissing(view(DataFrames._columns(df)[1],selected_rows))
end
function f3(df,selected_rows;disallowmissing=true)
    new_columns = Vector{AbstractVector}(undef, ncol(df))
    df_columns = df isa DataFrame ? DataFrames._columns(df) : collect(AbstractVector, eachcol(df))
    for i in eachindex(df_columns)
        if disallowmissing
            new_columns[i] = Missings.disallowmissing(view(df_columns[i],selected_rows))
        end
    end
    new_columns
end
basecase = (; rows_total=10^3, cols_total=10, cols_missing=3, missingness_share=0.1)
df = generate_data(; basecase...)
rowidxs = completecases(df)
selected_rows = DataFrames.findall(rowidxs)

# no failures -- all runs fine
f1(df,selected_rows)
f2(df,selected_rows)
f3(df,selected_rows)
dropmissing(df)

@code_lowered f1(df,selected_rows)
@code_lowered f2(df,selected_rows)
@code_llvm f1(df,selected_rows)
@code_llvm f2(df,selected_rows)
```

svilupp replied:
Oh, I'm slow. It was the same issue but with the other keyword -- we use keyword `view`, which overwrites the call to `view()`

I'll import it implicitly

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,52 @@ julia> dropmissing(df, [:x, :y])
+                end
+            end
+        end
+        # Output will be DataFrame even if SubDataFrame is provided
```
This is OK and intended.

bkamins commented:
Looks good. Can you please update tests to ensure full test coverage?

@nalimilan - maybe indeed everywhere we should reduce the multithreading threshold from `1_000_000` to `100_000` (I think it should be OK; @svilupp - I understand that this is consistent with your tests - right?).

bkamins commented:
When you add tests can you please check 0-column case? (I do not remember if we have it covered, but we should ensure it is, both: data frame with no columns, and data frame with some columns, but not no columns selected for dropping missing values).

bkamins commented:
Ah - and some note that this function uses multi-threading should be add in the docs (in particular in the section where we list functions using multi-threading).

svilupp commented:
> Looks good. Can you please update tests to ensure full test coverage?
> 
> @nalimilan - maybe indeed everywhere we should reduce the multithreading threshold from `1_000_000` to `100_000` (I think it should be OK; @svilupp - I understand that this is consistent with your tests - right?).

Yes, indeed. See the updated benchmark for the new `dropmissing` vs dropmissing that's always threaded.
On my machine, I can see the breakeven already at a 3-column dataframe, which I would agree is the vast majority of use-cases that I see (more like 10-50 in my world).
(this breakeven has shifted slightly as the `dropmissing` is now even faster, but the take away is the same)

<img width="998" alt="image" src="https://user-images.githubusercontent.com/49557684/213906522-f80abdbe-67b8-4ba8-ab92-1b735189464b.png">

svilupp commented:
> When you add tests can you please check 0-column case? (I do not remember if we have it covered, but we should ensure it is, both: data frame with no columns, and data frame with some columns, but not no columns selected for dropping missing values).

Done. 
I've also added some test cases:
- CategoricalArrays
- `view` kwarg - I noticed duplication with a set below (common to dropmissing+unique), but I'm testing all pathways of how it interacts with `disallowmissing` kwarg; and also its return type (which is tested separately for `dropmissing` overall, but not with `view=true`)
- `disallowmissing` kwarg (and its return type) + whether it correctly applies only to the required column (and leaves the type of others)

I couldn't find any test for the correctness of multi-threaded implementations. 
I saw file `test/multithreading.jl`, but that's testing the kwarg and not correctness. Is it because test runners are usually single-threaded, so they wouldn't trigger the pathway anyway?

In my case, it's a mirror copy of the non-threaded, so I'm not too concerned.

nalimilan commented on docs/src/lib/functions.md:
```diff
@@ -25,6 +25,8 @@ This is a list of operations that currently make use of multi-threading:
     is a custom function (i.e. not for standard reductions, which use
     optimized single-threaded methods).
+- `dropmissing` when the provided `DataFrame` has more than 1 column and `view=false` 
+  (subset-ing of individual columns is spawned in separate tasks).
```
```suggestion
  (subsetting of individual columns is spawned in separate tasks).
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,47 @@ julia> dropmissing(df, [:x, :y])
+        cols_inds = index(df)[cols]
+
+        # Threading decision rule borrowed from `_threaded_getindex`
+        if Threads.nthreads() > 1 && ncol(df) > 1 && length(selected_rows) >= 1_000_000
+            @sync for i in eachindex(new_columns)
```
AFAICT you can avoid duplicating code in two branches and instead do this:
```
        @spawn_or_run Threads.nthreads() > 1 && ncol(df) > 1 && length(selected_rows) >= 1_000_000 for i in eachindex(new_columns)
```

svilupp replied:
Great tip! I wasn't aware of this util.

But I'm not 100% on the semantics, based on its usage in `splitapplycombine.jl`, I think this is how it should be used in this context:

```
use_threads = Threads.nthreads() > 1 && ncol(df) > 1 && length(selected_rows) >= 100_000
@sync for i in eachindex(new_columns)
    @spawn_or_run use_threads if disallowmissing && (i in cols_inds)
        # implicit imports of functions due to name clash with keyword arguments!
        new_columns[i] = Missings.disallowmissing(Base.view(df_columns[i],selected_rows))
    else
        new_columns[i] = df_columns[i][selected_rows]
    end
end
```

Changes
- moved to inner side of the loop, because it dispatches the task itself

nalimilan commented on test/data.jl:
```diff
@@ -200,6 +216,26 @@ end
+    @test eltype(dropmissing!(copy(df), :b, disallowmissing=true).a) == Union{Int, Missing}
+
+    # CategoricalArrays
+    c = Union{Int64, Missing}[1,2,1,missing]|>categorical
```
```suggestion
    c = categorical([1, 2, 1, missing])
```

nalimilan commented on test/data.jl:
```diff
@@ -200,6 +216,26 @@ end
+    c = Union{Int64, Missing}[1,2,1,missing]|>categorical
+    df = DataFrame(c=c)
+    @test dropmissing(df) == DataFrame(c=categorical([1, 2, 1]))
+    @test eltype(dropmissing(df).c) == CategoricalValue{Int64, UInt32}
+    @test eltype(dropmissing!(df).c) == CategoricalValue{Int64, UInt32}
```
```suggestion
    @test eltype(dropmissing(df).c) == CategoricalValue{Int, UInt32}
    @test eltype(dropmissing!(df).c) == CategoricalValue{Int, UInt32}
```

nalimilan commented on test/data.jl:
```diff
@@ -200,6 +216,26 @@ end
     @test eltype(dropmissing(df).b) == Int
     @test eltype(dropmissing!(df).b) == Int
+
+    # disallowmissing=false
```
```suggestion
    # disallowmissing argument
```

nalimilan commented on test/data.jl:
```diff
@@ -186,6 +192,16 @@ end
+    @test dropmissing(df, view=true) == view(df, [1, 3], :)
+    @test typeof(dropmissing(df, view=true)) <: SubDataFrame
+    @test eltype(dropmissing(df, view=true, disallowmissing=false).a) == Union{Int, Missing}
+    # @test_throws ArgumentError dropmissing(df, view=true, disallowmissing=true) # tested separately
```
Why keep this commented out?

svilupp replied:
This exact path is tested slightly lower in the kwarg test shared by dropmissing and unique, so I didn’t want to duplicate it, but equally I didn’t want to delete it to show that it was considered. It takes milliseconds, so I should probably just keep it.

nalimilan commented on test/data.jl:
```diff
@@ -168,6 +168,12 @@ end
+    # Zero column case
+    isempty(dropmissing(DataFrame())) && dropmissing(DataFrame()) isa DataFrame
+    isempty(dropmissing!(DataFrame())) && dropmissing!(DataFrame()) isa DataFrame
+    @test_throws ArgumentError dropmissing(df1,[]) 
+    @test_throws ArgumentError dropmissing!(df1,[]) 
```
```suggestion
    @test_throws ArgumentError dropmissing(df1, []) 
    @test_throws ArgumentError dropmissing!(df1, []) 
```

nalimilan commented on src/other/utils.jl:
```diff
@@ -476,4 +476,4 @@ function _findall(B::BitVector)::Union{UnitRange{Int}, Vector{Int}}
     end
     @assert false "should not be reached"
-end
+end
```
```suggestion
end

```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,47 @@ julia> dropmissing(df, [:x, :y])
+                # for each column, check if disallowmissing should be applied
+                Threads.@spawn if disallowmissing && (i in cols_inds)
+                    # implicit imports of functions due to name clash with keyword arguments!
+                    new_columns[i] = Missings.disallowmissing(Base.view(df_columns[i],selected_rows))
```
```suggestion
                    new_columns[i] = Missings.disallowmissing(Base.view(df_columns[i], selected_rows))
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,47 @@ julia> dropmissing(df, [:x, :y])
+            @sync for i in eachindex(new_columns)
+                # for each column, check if disallowmissing should be applied
+                Threads.@spawn if disallowmissing && (i in cols_inds)
+                    # implicit imports of functions due to name clash with keyword arguments!
```
"Implicit" should be "explicit". But better drop the comment as this is a common pattern.
```suggestion
```

nalimilan commented:
> @nalimilan - maybe indeed everywhere we should reduce the multithreading threshold from 1_000_000 to 100_000 (I think it should be OK; @svilupp - I understand that this is consistent with your tests - right?).

Sounds reasonable. I don't think we benchmarked this thoroughly in when we chose the threshold -- we mainly wanted to be on the safe side. 

> I couldn't find any test for the correctness of multi-threaded implementations.
> I saw file test/multithreading.jl, but that's testing the kwarg and not correctness. Is it because test runners are usually single-threaded, so they wouldn't trigger the pathway anyway?
> In my case, it's a mirror copy of the non-threaded, so I'm not too concerned.

CI does use multithreading, so all that's needed is to test data frames which match the conditions that enable multithreading (number of rows, etc.). Tests should be added to the same place where `dropmissing` is tested (multithreading.jl is just about the keyword argument).

svilupp commented:
> CI does use multithreading, so all that's needed is to test data frames which match the conditions that enable multithreading (number of rows, etc.). Tests should be added to the same place where `dropmissing` is tested (multithreading.jl is just about the keyword argument).

I've added a test for the multithreaded path with a deterministic missing mask to make it stable.
I haven't verified all kwargs combinations, because that's tested elsewhere.
However, I've added a case for eltype when only specific columns are requested for `disallowmissing`, because that's inside of the threaded logic.

nalimilan commented on test/data.jl:
```diff
@@ -231,11 +231,31 @@ end
+    @test eltype(dropmissing!(df).c) == CategoricalValue{Int, UInt32}
+
+    # Multithreaded execution test (must be at least ncol > 1, nrow > 100_000)
+    N_rows, N_cols = 10^7, 4
```
Better not use more memory than necessary:
```suggestion
    N_rows, N_cols = 200_000_000, 4
```

svilupp replied:
I suspect you mean `200_000`, I'll reflect the change.

bkamins commented:
@svilupp - as a general comment - we do not need to aggressively prune duplicated tests. It is more important to keep all relevant tests in one place (and if they are relevant in two places have them run twice). It is much easier to ensure that we have 100% test coverage this way.

bkamins commented on docs/src/lib/functions.md:
```diff
@@ -25,6 +25,8 @@ This is a list of operations that currently make use of multi-threading:
   * a transformation produces one row per group and the passed transformation
     is a custom function (i.e. not for standard reductions, which use
     optimized single-threaded methods).
+- `dropmissing` when the provided `DataFrame` has more than 1 column and `view=false` 
```
```suggestion
- `dropmissing` when the provided data frame has more than 1 column and `view=false` 
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,35 @@ julia> dropmissing(df, [:x, :y])
     else
-        newdf = df[rowidxs, :]
-        disallowmissing && disallowmissing!(newdf, cols)
+        # Faster when there are many columns (Indexing with Integers than via Bool mask)
```
```suggestion
        # Faster when there are many columns (indexing with integers than via Bool mask)
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,35 @@ julia> dropmissing(df, [:x, :y])
+            end
+        end
+
+        newdf = DataFrame(new_columns, copy(DataFrames.index(df)), copycols=false)
```
```suggestion
        newdf = DataFrame(new_columns, copy(index(df)), copycols=false)
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,35 @@ julia> dropmissing(df, [:x, :y])
+        # Faster when there are many columns (indexing with integers than via Bool mask)
+        # or when there are many Missing (as we skip a lot of iterations)
+        selected_rows = _findall(rowidxs)
+        new_columns = Vector{AbstractVector}(undef, ncol(df))
+        df_columns = df isa DataFrame ? _columns(df) : collect(AbstractVector, eachcol(df))
+
+        # What column indices should disallowmissing be applied to
+        cols_inds = index(df)[cols]
+        
+        use_threads = Threads.nthreads() > 1 && ncol(df) > 1 && length(selected_rows) >= 100_000
+        @sync for i in eachindex(new_columns)
+            @spawn_or_run use_threads if disallowmissing && (i in cols_inds)
+                new_columns[i] = Missings.disallowmissing(Base.view(df_columns[i], selected_rows))
+            else
+                new_columns[i] = df_columns[i][selected_rows]
+            end
+        end
```
```suggestion
        new_columns = df isa DataFrame ? copy(_columns(df)) : collect(AbstractVector, eachcol(df))

        # What column indices should disallowmissing be applied to
        cols_inds = index(df)[cols]
        
        use_threads = Threads.nthreads() > 1 && ncol(df) > 1 && length(selected_rows) >= 100_000
        @sync for (i, col) in enumerate(new_columns)
            @spawn_or_run use_threads if disallowmissing && (i in cols_inds)
                new_columns[i] = Missings.disallowmissing(Base.view(col, selected_rows))
            else
                new_columns[i] = col[selected_rows]
            end
        end
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,34 @@ julia> dropmissing(df, [:x, :y])
+        cols_inds = index(df)[cols]
+        
+        use_threads = Threads.nthreads() > 1 && ncol(df) > 1 && length(selected_rows) >= 100_000
+        @sync for (i, col) in enumerate(new_columns)
```
```suggestion
        @sync for (i, col) in enumerate(eachcol(df))
```

bkamins commented on test/data.jl:
```diff
@@ -168,6 +168,12 @@ end
     end
 
+    # Zero column case
+    isempty(dropmissing(DataFrame())) && dropmissing(DataFrame()) isa DataFrame
+    isempty(dropmissing!(DataFrame())) && dropmissing!(DataFrame()) isa DataFrame
```
```suggestion
    @test isempty(dropmissing(DataFrame())) && dropmissing(DataFrame()) isa DataFrame
    @test isempty(dropmissing!(DataFrame())) && dropmissing!(DataFrame()) isa DataFrame
    df = DataFrame(a=1:3, b=4:6)
    dfv = @view df[:, 2:1]
    @test isempty(dropmissing(dfv) && dropmissing(dfv) isa DataFrame
    @test_throws ArgumentError dropmissing!(DataFrame())
```

bkamins commented on test/data.jl:
```diff
@@ -200,6 +216,46 @@ end
+    @test eltype(dropmissing!(df).c) == CategoricalValue{Int, UInt32}
+
+    # Multithreaded execution test (must be at least ncol > 1, nrow > 100_000)
+    N_rows, N_cols = 200_000, 3
```
```suggestion
    N_rows, N_cols = 110_000, 3
```

bkamins commented on test/data.jl:
```diff
@@ -200,6 +216,46 @@ end
+    # Deterministic drop mask: IF remainder of index position divided by 10 == column index THEN missing
+    for i in 1:ncol(df)
+        missing_mask = (eachindex(df[!,i]) .% 10) .== i
+        df[missing_mask,i] .= missing 
```
```suggestion
        df[missing_mask, i] .= missing 
```

bkamins commented:
@svilupp - note that I have pushed some changes to the PR.

bkamins commented on test/data.jl:
```diff
@@ -168,6 +168,16 @@ end
+    @test isempty(dropmissing!(DataFrame())) && dropmissing!(DataFrame()) isa DataFrame
+    df = DataFrame(a=1:3, b=4:6)
+    dfv = @view df[:, 2:1]
+    @test isempty(dropmissing(dfv) && dropmissing(dfv) isa DataFrame
```
```suggestion
    # TODO: re-enable after https://github.com/JuliaData/DataFrames.jl/issues/3272 is resolved
    # @test isempty(dropmissing(dfv)) && dropmissing(dfv) isa DataFrame
```

bkamins commented on test/data.jl:
```diff
@@ -168,6 +168,17 @@ end
+    dfv = @view df[:, 2:1]
+    # TODO: re-enable after https://github.com/JuliaData/DataFrames.jl/issues/3272 is resolved
+    # @test isempty(dropmissing(dfv)) && dropmissing(dfv) isa DataFrame
+    @test_throws ArgumentError dropmissing!(DataFrame())
```
```suggestion
    @test_throws ArgumentError dropmissing!(dfv)
```

bkamins commented:
@svilupp - can you please, when you have time, check the changes in the code I made (I have made them in my head, but they pass tests, so hopefully they are OK 😄). Thank you!

svilupp commented:
> @svilupp - can you please, when you have time, check the changes in the code I made (I have made them in my head, but they pass tests, so hopefully they are OK 😄). Thank you!

All looks good! I really like the change to `enumerate(eachcol(df))` - much more elegant.

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,34 @@ julia> dropmissing(df, [:x, :y])
+            else
+                new_columns[i] = col[selected_rows]
+            end
+        end
```
I wonder why coverage thinks this line isn't tested. I hope this doesn't mean the threaded path isn't covered.

bkamins replied:
I think it is a false positive. Note that the condition is not on line 996, but inside the loop.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,34 @@ julia> dropmissing(df, [:x, :y])
-        newdf = df[rowidxs, :]
-        disallowmissing && disallowmissing!(newdf, cols)
+        # Faster when there are many columns (indexing with integers than via Bool mask)
+        # or when there are many Missing (as we skip a lot of iterations)
```
```suggestion
        # or when there are few Missing (as we skip a lot of iterations)
```

svilupp replied:
I’m mot sure I follow your edit here, @bkamins 

My intuition is as follows:

Let’s take a vector of length 1000.
If we had no missings, findall returns 1:000 and we need to copy 1000 elements in the loop to the new vector.
If we had 999 missings, findall returns 1:1 and we are done after one loop iteration.
The above wouldn’t be the case for a Boolean mask, where we would need to check each one of the 1000 elements.

So shouldn’t the comment be reverted to previous form? ie, the more missings we have, the more iterations we can skip when copying the data?

bkamins replied:
You are right - I am not sure why I was confused (but it is good that we agree - I had the same in my mind 😄)

bkamins commented on test/data.jl:
```diff
@@ -200,6 +221,46 @@ end
+    end
+    
+    notmissing_rows = [i for i in 1:N_rows if i % 10 == 0 || i % 10 > ncol(df)]
+    @test isequal(dropmissing(df), df[notmissing_rows, :])
```
```suggestion
    @test dropmissing(df) ≅ df[notmissing_rows, :]
```

bkamins commented on test/data.jl:
```diff
@@ -200,6 +221,46 @@ end
+    cols = [:x1, :x2]
+    notmissing_rows = [i for i in 1:N_rows if i % 10 == 0 || i % 10 > length(cols)]
+    returned = dropmissing(df, cols)
+    @test isequal(returned, df[notmissing_rows, :])
```
```suggestion
    @test returned ≅ df[notmissing_rows, :]
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,34 @@ julia> dropmissing(df, [:x, :y])
-        newdf = df[rowidxs, :]
-        disallowmissing && disallowmissing!(newdf, cols)
+        # Faster when there are many columns (indexing with integers than via Bool mask)
+        # or when there are few Missing (as we skip a lot of iterations)
```
```suggestion
        # or when there are few missings (as we skip a lot of iterations)
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -970,15 +970,34 @@ julia> dropmissing(df, [:x, :y])
 @inline function dropmissing(df::AbstractDataFrame,
                              cols::Union{ColumnIndex, MultiColumnIndex}=:;
                              view::Bool=false, disallowmissing::Bool=!view)
+    # Identify Bool mask of which rows have no Missings
```
```suggestion
    # Identify Bool mask of which rows have no missings
```

nalimilan reviewed (approved):
Thanks!

bkamins commented:
Thank you! I hope you enjoyed the process. You are welcome to open other PRs.
